### PR TITLE
[Story image]

### PR DESCRIPTION
**Pull request by lfasani**
This commit depends 
* on the commits available on Sirius. It will not compile until the proper Sirius version is integrated
* on https://github.com/eclipse/kitalpha/pull/706

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    //@formatter:off
+    List<Resource> resources = resourceSet.getResources().stream()
+        .filter(res -> {
+          return CapellaResourceHelper.isCapellaResource(res) || new ResourceQuery(res).isAirdOrSrmResource(); 
+        }).filter(res -> {
+          //keep only the resource corresponding to the context otherwise the migration may be done many times for some resources.
+          IFile file = ResourcesPlugin.getWorkspace().getRoot().getFile(new org.eclipse.core.runtime.Path(res.getURI().toPlatformString(true)));
+          IFile contextFile = context.getResource();
+          return contextFile!=null && contextFile.equals(file);
+        }).collect(Collectors.toList());
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:80]: Not sure about this loop
resources seems always returning one resource if .capella or .aird/srm
`Resource resource = resourceSet.getResource(EcoreUtil2.getURI(context.getResource()), false); ` ?

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+      String nonCreatedFilesPath = nonCreatedFiles.stream().map(File::getAbsolutePath)
+          .collect(Collectors.joining(", "));
+      Activator.getDefault().getLog()
+          .warn(MessageFormat.format(Messages.MigrationAction_Image_AsolutePathImageNotMigrated,
+              new EditingDomainServices().getLabelProviderText(notifier), attribute.getName(), nonCreatedFilesPath));
+    }
+    return createdFiles;
+  }
+
+  private IProject getProject(Resource resource) {
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:219]: EcoreUtil2.getFile(resource).getProject()

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    return iFile.getProject();
+  }
+
+  private IFolder getImageFolder(EObject contextObject) {
+    String platformString = contextObject.eResource().getURI().toPlatformString(true);
+
+    IFile airdFile = ResourcesPlugin.getWorkspace().getRoot()
+        .getFile(new org.eclipse.core.runtime.Path(platformString));
+
+    return airdFile.getProject().getFolder(ImageManager.IMAGE_FOLDER_NAME);
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:233]: EcoreUtil2.getProject(contextObject).getFolder(ImageManager.IMAGE_FOLDER_NAME);

**Review by pdulth — CHANGES_REQUESTED** [review]

File: core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java [hunk 4]
```diff
@@ -141,7 +146,74 @@ public void run() {
...
+
+    updateRichTextString(session, pm);
+
+  }
+
+  private void updateRichTextString(Session session, IProgressMonitor pm) {
+    pm.subTask(org.polarsys.capella.core.sirius.ui.internal.Messages.updating_richtext_description);
+    Collection<Resource> semanticResources = session.getSemanticResources();
+    for (Resource resource : semanticResources) {
+      if (CapellaResourceHelper.isCapellaResource(resource)) {
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java:162]: Description on diagrams?

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+            .error(MessageFormat.format(Messages.MigrationAction_Image_ImpossibleToCreateImageFolder,
+                imageFolder.getFullPath(), new EditingDomainServices().getLabelProviderText(contextObject),
+                attribute.getName(), nonCreatedFilesPath), e);
+      }
+    }
+
+    // TODO attention aux images project relative qui serait vu comme des images absolute sur linux
+    // -> non car si le path exist c'est donc qu'il est absolu
+    for (File fileToCopy : filesToCopy.keySet()) {
+      // FileOutputStream fileOutputStream;
```
[the remark this answers is not in the document]
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:275]: done

File: core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java [hunk 6]
```diff
@@ -141,7 +146,74 @@ public void run() {
...
+        TreeIterator<EObject> allContents = resource.getAllContents();
+        while (allContents.hasNext()) {
+          EObject eObject = allContents.next();
+
+          List<EStructuralFeature> featuresToMigrate = eObject.eClass().getEAllStructuralFeatures().stream()
```
[the remark this answers is not in the document]
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java:167]: done

File: core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java [hunk 7]
```diff
@@ -141,7 +146,74 @@ public void run() {
...
+                eObject.eSet(attr, textWithUpdatedImagePaths);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private void updatWorkspaceImage(Session session, IProgressMonitor pm) {
```
[the remark this answers is not in the document]
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java:204]: done

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        }).collect(Collectors.toList());
+    //@formatter:on
+
+    Set<EAttribute> richTextAttributes = RichTextAttributeRegistry.INSTANCE.getEAttributes();
+    for (Resource resource : resources) {
+      TreeIterator<EObject> allContents = resource.getAllContents();
+      while (allContents.hasNext()) {
+        EObject eObject = allContents.next();
+        // keep only EAttrutes that are declared as containing html content
+        List<EStructuralFeature> featuresToMigrate = eObject.eClass().getEAllStructuralFeatures().stream()
```
[the remark this answers is not in the document]
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:89]: done

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java (hunk 2 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:219]: done

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java (hunk 3 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java:233]: done

File: core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java (hunk 4 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.sirius.ui/src/org/polarsys/capella/core/sirius/ui/refactoring/WorkspaceImagePathChange.java:162]: indeed!

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java (hunk 1 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/contribution/ImagePathInRichTextAttributeContribution.java]: done

**Comment by pdulth** [comment]
Replace non-opensource JoyPixels dinosaur.png with EPL-2.0 capella.png one on junit model

**Pull request merged**